### ISSUE TITLE
Two objects with the same property but distinct literal types cannot be assigned to their union type

### ISSUE THREAD

**Issue by niedzielski** [Microsoft/TypeScript#30385]
**TypeScript Version:**  v3.4.0-dev.20190312

**Search Terms:** deep literal type union equivalence

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.

interface Mouse {
  readonly type: 'mouse'
}

interface Joystick {
  readonly type: 'joystick'
}

// OK
function foo(input: number): Mouse | Joystick {
  const type: 'mouse' | 'joystick' = input ? 'mouse' : 'joystick'
  return type === 'mouse' ? /*: { type: "mouse"; } */ {type} : /*: { type: "joystick"; } */ {type} // Valid.
}

// Unexpected error
function bar(input: number): Mouse | Joystick {
  const type: 'mouse' | 'joystick' = input ? 'mouse' : 'joystick'
  return {type} // Unexpectedly invalid: Type '"mouse"' is not assignable to type '"joystick"'.
  /* { type: "mouse" | "joystick"; } instead of { type: "mouse"; } | { type: "joystick"; } */
}

type Keyboard = {
  readonly type: 'keyboard'
}

type Touchscreen = {
  readonly type: 'touchscreen'
}

// Unexpected error (with types instead of interfaces)
function baz(input: number): Keyboard | Touchscreen {
  const type: 'keyboard' | 'touchscreen' = input ? 'keyboard' : 'touchscreen'
  return {type} // Unexpectedly invalid: Type '"keyboard"' is not assignable to type '"touchscreen"'.
  /* { type: "keyboard" | "touchscreen"; } instead of { type: "keyboard"; } | { type: "touchscreen"; } */
}

// Expected error
function boo(input: number): Keyboard | Touchscreen {
  const ret /*: { type: string; } */ = input ? { type: 'keyboard' } : { type: 'touchscreen' }
  return ret // Correctly invalid: Type 'string' is not assignable to type '"touchscreen"'.
}
```

**Expected behavior:**
TypeScript infers the union of the top level type.

**Actual behavior:**
TypeScript infers the union of the subtypes.

[**Playground Link**](https://www.typescriptlang.org/play/#src=interface%20Mouse%20%7B%0D%0A%20%20readonly%20type%3A%20'mouse'%0D%0A%7D%0D%0A%0D%0Ainterface%20Joystick%20%7B%0D%0A%20%20readonly%20type%3A%20'joystick'%0D%0A%7D%0D%0A%0D%0Afunction%20foo(input%3A%20number)%3A%20Mouse%20%7C%20Joystick%20%7B%0D%0A%20%20const%20type%3A%20'mouse'%20%7C%20'joystick'%20%3D%20input%20%3F%20'mouse'%20%3A%20'joystick'%0D%0A%20%20return%20type%20%3D%3D%3D%20'mouse'%20%3F%20%2F*%3A%20%7B%20type%3A%20%22mouse%22%3B%20%7D%20*%2F%20%7Btype%7D%20%3A%20%2F*%3A%20%7B%20type%3A%20%22joystick%22%3B%20%7D%20*%2F%20%7Btype%7D%20%2F%2F%20Valid.%0D%0A%7D%0D%0A%0D%0Afunction%20bar(input%3A%20number)%3A%20Mouse%20%7C%20Joystick%20%7B%0D%0A%20%20const%20type%3A%20'mouse'%20%7C%20'joystick'%20%3D%20input%20%3F%20'mouse'%20%3A%20'joystick'%0D%0A%20%20return%20%7Btype%7D%20%2F%2F%20Unexpectedly%20invalid%3A%20Type%20'%22mouse%22'%20is%20not%20assignable%20to%20type%20'%22joystick%22'.%0D%0A%20%20%2F*%20%7B%20type%3A%20%22mouse%22%20%7C%20%22joystick%22%3B%20%7D%20instead%20of%20%7B%20type%3A%20%22mouse%22%3B%20%7D%20%7C%20%7B%20type%3A%20%22joystick%22%3B%20%7D%20*%2F%0D%0A%7D%0D%0A%0D%0Atype%20Keyboard%20%3D%20%7B%0D%0A%20%20readonly%20type%3A%20'keyboard'%0D%0A%7D%0D%0A%0D%0Atype%20Touchscreen%20%3D%20%7B%0D%0A%20%20readonly%20type%3A%20'touchscreen'%0D%0A%7D%0D%0A%0D%0Afunction%20baz(input%3A%20number)%3A%20Keyboard%20%7C%20Touchscreen%20%7B%0D%0A%20%20const%20type%3A%20'keyboard'%20%7C%20'touchscreen'%20%3D%20input%20%3F%20'keyboard'%20%3A%20'touchscreen'%0D%0A%20%20return%20%7Btype%7D%20%2F%2F%20Unexpectedly%20invalid%3A%20Type%20'%22keyboard%22'%20is%20not%20assignable%20to%20type%20'%22touchscreen%22'.%0D%0A%20%20%2F*%20%7B%20type%3A%20%22keyboard%22%20%7C%20%22touchscreen%22%3B%20%7D%20instead%20of%20%7B%20type%3A%20%22keyboard%22%3B%20%7D%20%7C%20%7B%20type%3A%20%22touchscreen%22%3B%20%7D%20*%2F%0D%0A%7D%0D%0A%0D%0Afunction%20boo(input%3A%20number)%3A%20Keyboard%20%7C%20Touchscreen%20%7B%0D%0A%20%20const%20ret%20%2F*%3A%20%7B%20type%3A%20string%3B%20%7D%20*%2F%20%3D%20input%20%3F%20%7B%20type%3A%20'keyboard'%20%7D%20%3A%20%7B%20type%3A%20'touchscreen'%20%7D%0D%0A%20%20return%20ret%20%2F%2F%20Correctly%20invalid%3A%20Type%20'string'%20is%20not%20assignable%20to%20type%20'%22touchscreen%22'.%0D%0A%7D%0D%0A) 

**Related Issues:** 

I am certain this issue has been reported previously and apologize that I was unable to identify it.

**Comment by jack-williams** [comment]
Duplicate of #8289

**Comment by niedzielski** [comment]
Thank you @jack-williams. In this example, the suggested approach is something like: 

```ts
interface Device {
  readonly type: Mouse['type'] | Joystick['type']
}
```
